src/Graphics/Image/Processing/Ahe.hs

```diff
@@ -0,0 +1,85 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE BangPatterns #-} 
+
+-- | Adaptive histogram equalization is used to improve contrast in images.
+-- It adjusts image intensity in small regions (neighborhood) in the image.
+module Graphics.Image.Processing.Ahe where
+
+import Control.Monad (forM_, when)
+import Control.Monad.ST
+import Data.STRef 
+import Debug.Trace (trace)
+
+import Prelude as P hiding (subtract)
+import Graphics.Image.Processing.Filter
+import Graphics.Image.Interface as I
+import Graphics.Image
+import Graphics.Image.Types as IP
+import Graphics.Image.ColorSpace (X)
+
+-- | Supplementary function for applying border resolution and a general mask.
+simpleFilter :: (Array arr cs e, Array arr X e) => Direction -> Border (Pixel cs e) -> Filter arr cs e
+simpleFilter dir !border =
+  Filter (correlate border kernel)
+  where
+    !kernel =
+      case dir of
+        Vertical   -> fromLists $ [ [ 0, -1, 0 ], [  -1,  4,  -1 ], [  0,  -1,  0 ] ]
+        Horizontal -> fromLists $ [ [ 0, -1, 0 ], [ -1, 4, -1 ], [ 0, -1, 0 ] ]
+
+-- | 'ahe' operates on small 'contextual' regions of the image. It enhances the contrast of each
+-- region and this technique works well when the distribution of pixel values is similar throughout
+-- the image. 
+--
+-- The idea is to perform contrast enhancement in 'neighborhood region' of each pixel and the size
+-- of the region is a parameter of the method. It constitutes a characteristic length scale: contrast 
+-- at smaller scales is enhanced, while contrast at larger scales is reduced (For general purposes, a size
+-- factor of 5 tends to give pretty good results).
+--
+-- <<images/yield.jpg>>   <<images/yield_ahe.png>>
+--
+-- Usage : 
+--	
+-- >>> img <- readImageY VU "yield.jpg"
+-- >>> input1 <- getLine
+-- >>> input2 <- getLine
+-- >>> let thetaSz = (P.read input1 :: Int)
+-- >>> let distSz = (P.read input2 :: Int) 
+-- >>> let neighborhoodFactor = (P.read input2 :: Int) 
+-- >>> let aheImage :: Image VU RGB Double
+-- >>>     aheImage = ahe img thetaSz distSz neighborhoodFactor
+-- >>> writeImage "test.png" (toImageRGB aheImage)
```

Comment:
And this should be `"images/yield_ahe.png"`, since that's what it is called in the repo.

Reply:
Yes sir, very sorry. Will change it.